src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs

```diff
@@ -79,78 +66,15 @@ private static MethodInfo GetGenericMethodMethodInfo()
             Debug.Assert(baseType != null);
             Debug.Assert(interfaceType != null);
 
-            GeneratedTypeInfo proxiedType = GetProxyType(baseType, interfaceType);
-            return Activator.CreateInstance(proxiedType.GeneratedType, new object[] { proxiedType.MethodInfos })!;
-        }
-
-        private static GeneratedTypeInfo GetProxyType(
-            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] Type baseType,
-            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type interfaceType)
-        {
-            lock (s_baseTypeAndInterfaceToGeneratedProxyType)
-            {
-                if (!s_baseTypeAndInterfaceToGeneratedProxyType.TryGetValue(baseType, out Dictionary<Type, GeneratedTypeInfo>? interfaceToProxy))
-                {
-                    interfaceToProxy = new Dictionary<Type, GeneratedTypeInfo>();
-                    s_baseTypeAndInterfaceToGeneratedProxyType[baseType] = interfaceToProxy;
-                }
-
-                if (!interfaceToProxy.TryGetValue(interfaceType, out GeneratedTypeInfo? generatedProxy))
-                {
-                    generatedProxy = GenerateProxyType(baseType, interfaceType);
-                    interfaceToProxy[interfaceType] = generatedProxy;
-                }
-
-                return generatedProxy;
-            }
-        }
-
-        // Unconditionally generates a new proxy type derived from 'baseType' and implements 'interfaceType'
-        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2062:UnrecognizedReflectionPattern",
-            Justification = "interfaceType is annotated as preserve All members, so any Types returned from GetInterfaces should be preserved as well once https://github.com/mono/linker/issues/1731 is fixed.")]
-        private static GeneratedTypeInfo GenerateProxyType(
-            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] Type baseType,
-            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type interfaceType)
-        {
-            // Parameter validation is deferred until the point we need to create the proxy.
-            // This prevents unnecessary overhead revalidating cached proxy types.
-
-            // The interface type must be an interface, not a class
-            if (!interfaceType.IsInterface)
+            AssemblyLoadContext? alc = AssemblyLoadContext.GetLoadContext(baseType.Assembly);
```

Comment:
Thank you @teo-tsirpanis overall the fix LGTM

@vitek-karas [mentioned](https://github.com/dotnet/runtime/issues/60468#issuecomment-975968918) that we might want to discuss `if it should always use the ALC of the base type, or if it should react to the current contextual reflection setting. Right now I think it should probably react to contextual reflection, but it will probably need more thought.` I am not familiar with contextual reflection so I have no opinion, i would follow the expert's suggestions here

Reply:
I spent some more time on this - created https://github.com/dotnet/runtime/issues/62234 to track the `TypeBuilder` problem.

Now I think it's probably better to not react to contextual reflection. If we did use contextual reflection, it would be possible to hit the `TypeBuilder` problem. If we always rely on the base type's ALC that will never happen.
Note that I don't think we will ever really fix the `TypeBuilder` problem, the best we can do without massive rewrite is to actually outright fail creating the type in the "failure" case. Which would mean that using contextual reflection would lead to cases where the proxy creating fails. Not great.

It would also mean that it would be possible to create two proxy types for the exact same combination of base type and interface type (in two different ALCs) - kind of against the current behavior. And there could also be subtle issues if the custom ALC specified via contextual reflection have different assembly resolution behavior.

It still possible to get to weird behavior cases, like loading the same `IFace` type into two different ALCs, and then trying to create proxies using the same `BaseType` but over the two different `IFace` types. But honestly I don't know if this has a solution - either way I can see this scenario to break, so picking one is probably as good as the other.

Reply:
Thank you for looking into this